packages/react-native-gesture-handler/src/v3/detectors/GestureDetector.tsx

```diff
@@ -0,0 +1,25 @@
+import { NativeDetectorProps } from './common';
+import { BaseGesture } from '../../handlers/gestures/gesture';
+import { NativeDetector } from './NativeDetector';
+import { ComposedGesture } from '../../handlers/gestures/gestureComposition';
+import {
+  GestureDetectorProps as LegacyGestureDetectorProps,
+  GestureDetector as LegacyGestureDetector,
+} from '../../handlers/gestures/GestureDetector';
+import { useDetectorContext } from './LogicDetector/useDetectorContext';
+import { LogicDetector } from './LogicDetector/LogicDetector';
+export function GestureDetector<THandlerData, TConfig>(
+  props: NativeDetectorProps<THandlerData, TConfig> | LegacyGestureDetectorProps
+) {
+  const context = useDetectorContext();
+  return props.gesture instanceof ComposedGesture ||
+    props.gesture instanceof BaseGesture ? (
+    <LegacyGestureDetector {...(props as LegacyGestureDetectorProps)} />
+  ) : context ? (
+    <LogicDetector {...(props as NativeDetectorProps<THandlerData, TConfig>)} />
+  ) : (
+    <NativeDetector
+      {...(props as NativeDetectorProps<THandlerData, TConfig>)}
+    />
+  );
```

Comment:
```suggestion
  if (props.gesture instanceof ComposedGesture ||
    props.gesture instanceof BaseGesture) {
    return <LegacyGestureDetector {...(props as LegacyGestureDetectorProps)} />;
  }

  const context = use(DetectorContext);
  return context ? (
    <LogicDetector {...(props as NativeDetectorProps<THandlerData, TConfig>)} />
  ) : (
    <NativeDetector
      {...(props as NativeDetectorProps<THandlerData, TConfig>)}
    />
  );
```

If this works, then there's no need to read context when rendering the legacy detector (not sure if syntax is correct, but you get the idea).